### DISCUSSION TITLE
Test failures in main?

### DISCUSSION THREAD

**Discussion by gwhitney** [pnpm/pnpm#5301] [Q&A]
I've cloned the repository and compiled and "pd" appears to be working fine and main is up to date with the official repository and is what I have checked out with a clean working tree. But I have two problems with `pnpm test-main`:

1. If I run it twice, the second time it dies with `fatal--- cannot create server: listen EADDRINUSE: address already in use ::1:4873` in 
```
@pnpm/core@5.11.0 _test: `cross-env PNPM_REGISTRY_MOCK_PORT=4873 pnpm run test:e2e`
```
and so far the only way I have to get it back to running through that point is to reboot.
2. On that first run after a reboot, it still reports a failure in 
```
@pnpm/default-reporter@9.1.20 _test: `jest`
```
specifically in `test/reportingUpdateCheck.ts`, showing a diff between "Snapshot - 1" and "Received + 1" of:
```
│     -    │                Run "pnpm add -g pnpm" to update.                 
│     +    │     Run "corepack prepare pnpm@11.0.0 --activate" to update.
```
The fact is, I did originally install pnpm via corepack. So I suspect this is not a real failure, but it would be nice to have more certainty.

Any suggestions for how I can get clean test runs in main, especially when I run it repeatedly, would be very appreciated so that I can be sure any changes I may try are passing all tests. Thanks.

**Comment by zkochan** [comment]
> If I run it twice, the second time it dies with fatal--- cannot create server: listen EADDRINUSE: address already in use ::1:4873 in

I rarely get such issue. But you can fix it by killing the process:

```
pnpm dlx fkill-cli :4873
```

**Reply by gwhitney** [reply]
Great, thanks, that did take care of the first problem (only being able to run `pnpm test-main` once per reboot). Any thoughts on the other problem, how to make the tests robust to whatever method was used to install pnpm? I am willing to file a PR, but would need some guidance on how the test should be changed so that it will pass even though I originally installed pnpm with `corepack` (which is what I assume is causing the failure, based on the message). Thanks for letting me know your thoughts.

**Reply by zkochan** [reply]
I think I have fixed this particular issue: https://github.com/pnpm/pnpm/commit/32814aa96f8fc4aa82111c18bfee0ba5906b1a88

**Reply by gwhitney** [reply]
Yes, I pulled and it runs through that point. In fact, it runs much farther, but then it gets to a point where it has displayed the following and then just hangs (at least for half an hour) with no further output but without my prompt coming back (this is on Linux, again just the code in main with the latest commit). Is there something I am doing wrong?
```
packages/which-version-is-pinned _test$ jest
[7 lines collapsed]
│ File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
│ -83-------|---------|----------|---------|---------|-------------------
│   ● a package that need authentication reuses authorization tokens for tarba
│     GET http://127.0.0.1:4873/@pnpm.e2e%2Fneeds-auth: request to http://127.
│       60 |         }) as RegistryResponse
│       61 |       } catch (error: any) { // eslint-disable-line
│     > 62 |         reject(new PnpmError('META_FETCH_FAIL', `GET ${uri}: ${er
│          |                ^
│       63 |         return
│       64 |       }
│       65 |       if (response.status > 400) {
│       at RetryOperation._fn (../npm-resolver/src/fetch.ts:62:16)
[3 lines collapsed]
│ File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
│ ----------|---------|----------|---------|---------|-------------------
│ All files |     100 |    86.66 |     100 |     100 |                   
│  index.js |     100 |    86.66 |     100 |     100 | 3,21              
│ ----------|---------|----------|---------|---------|-------------------
│ Test Suites: 1 passed, 1 total
│ Tests:       1 passed, 1 total
│ Snapshots:   0 total
│ Time:        1.444 s, estimated 2 s
│ Ran all test suites.
└─ Done in 2.2s
```
Thanks for all of your help and attention to my issues!

**Reply by zkochan** [reply]
I don't have any idea why this one happens. Try to skip this particular test with `test.skip(...`

**Reply by gwhitney** [reply]
OK, so first I tried `pnpm '--filter=!which-version-is-pinned' test` and that errored after a while, so I tracked it down to `pnpm --filter=assert-store test` which produces the following output. This is on openSUSE Tumbleweed, up to date, with exactly the code in the latest commit of main:
```

> @pnpm/assert-store@1.0.46 test /home/glen/code/pnpm/privatePackages/assert-store
> pnpm pretest && pnpm run compile && jest


> @pnpm/assert-store@1.0.46 pretest /home/glen/code/pnpm/privatePackages/assert-store
> pnpm install -C test/fixture/project --force --no-shared-workspace-lockfile

 WARN  using --force I sure hope you know what you are doing
Lockfile is up to date, resolution step is skipped
Packages: +1
+
Progress: resolved 1, reused 1, downloaded 0, added 1, done

> @pnpm/assert-store@1.0.46 compile /home/glen/code/pnpm/privatePackages/assert-store
> tsc --build

 FAIL  test/index.ts
  ✕ assertStore() store assertions (2 ms)
  ✓ assertStore() resolve (1 ms)

  ● assertStore() store assertions

    expect(received).toBeTruthy()

    Received: false

       9 | ) => {
      10 |   // eslint-disable-next-line
    > 11 |   const ok = (value: any) => expect(value).toBeTruthy()
         |                                            ^
      12 |   // eslint-disable-next-line
      13 |   const notOk = (value: any) => expect(value).toBeFalsy()
      14 |   const ern = encodedRegistryName ?? `localhost+${REGISTRY_MOCK_PORT}`

      at toBeTruthy (src/index.ts:11:44)
      at Object.ok [as storeHas] (src/index.ts:31:7)
      at Object.<anonymous> (test/index.ts:10:3)

----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------|---------|----------|---------|---------|-------------------
All files |   63.33 |       60 |      50 |   64.28 |                   
 index.js |   63.33 |       60 |      50 |   64.28 | 18-28,35-36,41    
----------|---------|----------|---------|---------|-------------------
Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 passed, 2 total
Snapshots:   0 total
Time:        1.356 s, estimated 2 s
Ran all test suites.
/home/glen/code/pnpm/privatePackages/assert-store:
 ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  @pnpm/assert-store@1.0.46 test: `pnpm pretest && pnpm run compile && jest`
Exit status 1
```
Sorry to be bothering you with all this, would just be nice to get a clean test in main so I can be confident that any changes I might try are not detrimental. Thanks again for all the help you've provided.

**Reply by zkochan** [reply]
Wow, seems like all those storeHas assertions are broken for a long time. Now I don't even remember what they should check. Seems like they were broken when pnpm switched to a content-addressable store. I created an issue for now: https://github.com/pnpm/pnpm/issues/5334
